Feature - jwt 재발급 API 연결

yangsubinn opened the pull request:
## 관련 이슈
<!--
이슈 번호 #000 작성  
ex) close #1 
--> 
close #222 

## 내용
<!--
로직 설명  
--> 
1. UserAPI의 reissue, RefreshService 구현
2. Task+ retrying 함수 구현
  - 최초 서버 통신 실패시 한 번 더 시도하도록 하는 retrying 함수 추가했습니다.
  - 최초 시도에서 성공할 경우 그대로 넘어가게 됩니다.
  - 기존 request와 requestWithNoResult에서 retrying 내부에서 서버 통신 요청하도록 수정했습니다.
3. handleError400의 401 부분 수정
  - 401을 받은 경우는 전부 토큰 문제로 확인해서 토큰 재발급 요청을 보내도록 했습니다.

**로직**
1. Task.retrying 내부에서 최초로 서버 통신을 시도했을때 statusCode 401을 받으면
2. handleError400에서 토큰 재발급 요청을 보내게 됩니다.
3. 토큰 재발급이 정상적으로 된 경우(success), 저장되어있던 accessToken과 refreshToken이 새로 저장되고, retrying 내부에 있기 때문에 새롭게 갱신된 토큰으로 다시 한번 더 시도하게 됩니다.
4. 토큰 재발급이 실패할 경우(fail, keyChainError) unauthorized로 돌아가며 로그인으로 이동시킵니다. (각 뷰에서 로그인으로 이동 처리)


### 이미지, 영상
<!--
필요시 스크린샷 첨부
--> 
재시도 X

https://github.com/dnd-side-project/dnd-9th-2-ios/assets/81167570/d575f3a4-15d3-43b5-9ead-4b36d38bb5ec

재시도 O (통신 2회 요청)

https://github.com/dnd-side-project/dnd-9th-2-ios/assets/81167570/f82a5ea6-c371-442f-88c8-91c478a9f08c

## 리뷰어가 확인할 사항
<!--
중점적으로 봐주면 좋을 사항  
ex) ㅇㅇㅇㅇ하려고 Combine 사용했는데 제대로 사용하고 있는건가요?
--> 

전체적으로 모든 api에서 수정되어야 하는거라 작업은 안했는데,
위에 로직 3번에서 retrying 할때 새로 갱신된 토큰을 사용해서 요청 보내기 위해서는 token 넣어주는 부분이 현재 위치한 각 Service 부분이 아니라 API 자체 내부에 있어야 할 것 같습니다.

전체 로직이 잘 굴러갈지, 맞는지 확인을 못해서 우선 로직이 괜찮을지 확인해주시면 
token 부분을 Service에서 인자로 넘겨주지 않고, API에서 바로 가져다 쓰도록 옮기는 작업을 한번에 싹 해야될 것 같습니다 


## 기타
<!--
레퍼런스 혹은 패키지 설치 등
-->

https://stackoverflow.com/questions/75081648/swift-async-networking-how-do-i-retry-on-error
https://green1229.tistory.com/335

yangsubinn committed:
Feat #222 - 토큰 재발급 reissue UserAPI 연결

yangsubinn committed:
Feat #222 - RefreshService 구현

- RefreshEntity, RefreshService, RefreshServiceResult 생성

yangsubinn committed:
Feat #222 - Task+ retrying 함수 추가

yangsubinn committed:
Fix #222 - 네트워크 요청 최소 2번 시도하도록 NetworkService 수정

yangsubinn committed:
Chore #222 - 불필요한 코드 삭제

never-better commented on Baggle/Baggle/Core/Models/Network/Refresh/RefreshEntity.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct RefreshEntity: Codable {
```
p4: Token Refresh가 더 직관적일 것 같아요.

never-better commented on Baggle/Baggle/Core/Common/Extensions/Foundation/Task+.swift:
```diff
@@ -13,3 +13,30 @@ extension Task where Success == Never, Failure == Never {
+                    return try await operation()
+                } catch {
+                    print("✅ retrying")
+                    let oneSecond = TimeInterval(1_000_000_000)
```
p2: 11번째 줄에 있는 sleep함수를 재사용하면 코드를 줄일 수 있을 것 같아요.

never-better commented on Baggle/Baggle/Core/Services/Common/RefreshService.swift:
```diff
@@ -0,0 +1,47 @@
+import ComposableArchitecture
+import Moya
+
+struct RefreshService {
```
p3: Token Refresh가 직관적일 것 같아요

never-better commented on Baggle/Baggle/Core/Services/Common/NetworkService.swift:
```diff
@@ -48,66 +51,74 @@ extension NetworkService {
-        case 200, 201:
-            guard let data = body.data else {
-                throw APIError.unwrapping
+        try await _Concurrency.Task.retrying {
```
Q: `_Concurrency`는 왜 있는건가요??

yangsubinn replied:
그냥 Task로 하면 Moya의 Task가 연결돼서 지정해줬습니다

never-better reviewed (commented):
확인했습니다! 

저도 이거 이슈 파면서 각 Service마다 다 일일이 재발급 설정해줘야 하는데 어떻게 해야하나 걱정됐습니다...
제가 생각한 방법은 2가지이고, 둘 다 장단점이 있는 것 같습니다.

첫 번째는 수빈님이 작업하신 것처럼 Network Service 내부에서 401 에러 발생시 refresh를 해줘 처리해주는 방식입니다.
- 장점은 공통으로 처리하기 때문에 코드가 확실히 적다는 것입니다. 
- 단점은 네트워크 서비스(=Network) 안에 네트워크 서비스(=refresh)가 있고, 토큰 저장을 위해서 유저 매니저를 호출하게 돼 유저매니저에 종속된다는 것입니다. 프로토콜을 사용해서 의존성 역전을 해줄 수 있지만, 네트워크 서비스의 역할이 조금은 커지는 것 같습니다.

단점을 길게 쓰긴 했는데 그래도 한 번에 관리할 수 있다는 장점이 압도적으로 큰 것 같습니다.

두 번째는 각 서비스에서 토큰 에러를 캐치해 토큰 리프레쉬를 하고, 다시 새로운 토큰으로 재요청 하는 것입니다.
- 장점은 각 네트워크 서비스와 유저 매니저의 역할을 분리할 수 있다는 것입니다. TCA `Service` = 클린 아키텍처 `UseCase`라고 생각해서, 서비스 내부 로직을 통일성 있게 처리해 코드의 역할을 명확히 해주는 것 같습니다.
- 단점은 각 서비스마다 토큰 리프레쉬를 처리해줘야 하기 때문에 코드 양이 조금(?) 많아지는 것입니다.

![IMG_4435](https://github.com/dnd-side-project/dnd-9th-2-ios/assets/71776532/1d74829e-1c99-4da7-bbc1-236972e05382)

코드 구조를 생각하면 저는 2번으로 마음이 기울긴 하는데 수빈님 생각 말씀해주세요.

수고하셨습니다!

yangsubinn committed:
Fix #240 - 코드리뷰 반영

yangsubinn commented:
> 확인했습니다!
> 
> 저도 이거 이슈 파면서 각 Service마다 다 일일이 재발급 설정해줘야 하는데 어떻게 해야하나 걱정됐습니다... 제가 생각한 방법은 2가지이고, 둘 다 장단점이 있는 것 같습니다.
> 
> 첫 번째는 수빈님이 작업하신 것처럼 Network Service 내부에서 401 에러 발생시 refresh를 해줘 처리해주는 방식입니다.
> 
> * 장점은 공통으로 처리하기 때문에 코드가 확실히 적다는 것입니다.
> * 단점은 네트워크 서비스(=Network) 안에 네트워크 서비스(=refresh)가 있고, 토큰 저장을 위해서 유저 매니저를 호출하게 돼 유저매니저에 종속된다는 것입니다. 프로토콜을 사용해서 의존성 역전을 해줄 수 있지만, 네트워크 서비스의 역할이 조금은 커지는 것 같습니다.
> 
> 단점을 길게 쓰긴 했는데 그래도 한 번에 관리할 수 있다는 장점이 압도적으로 큰 것 같습니다.
> 
> 두 번째는 각 서비스에서 토큰 에러를 캐치해 토큰 리프레쉬를 하고, 다시 새로운 토큰으로 재요청 하는 것입니다.
> 
> * 장점은 각 네트워크 서비스와 유저 매니저의 역할을 분리할 수 있다는 것입니다. TCA `Service` = 클린 아키텍처 `UseCase`라고 생각해서, 서비스 내부 로직을 통일성 있게 처리해 코드의 역할을 명확히 해주는 것 같습니다.
> * 단점은 각 서비스마다 토큰 리프레쉬를 처리해줘야 하기 때문에 코드 양이 조금(?) 많아지는 것입니다.
> 
> ![IMG_4435](https://user-images.githubusercontent.com/71776532/265938884-1d74829e-1c99-4da7-bbc1-236972e05382.jpg)
> 
> 코드 구조를 생각하면 저는 2번으로 마음이 기울긴 하는데 수빈님 생각 말씀해주세요.
> 
> 수고하셨습니다!

저도 처음에 두번째를 생각했어서, 유탁님 말씀에 동의합니다 !
다만 두번째로 하게되면 말씀하신 것처럼 
기존에 있던 모든 service에서 중복적으로 처리를 해줘야 하기도 하고, 토큰이 없는 경우까지 각 service들이 알고 직접 다시 처리해줘야 할까? 하는 생각에 우선은 첫번째 방법으로 했던 것 같습니다 .

일단은 2번이 명확하게 역할 분배가 된 구조로 더 적합한 것 같네요..
근데 토큰 에러를 catch해서 refresh하고 재요청 보내는 과정의 구현이 간단하지는 않을 것 같아서, 해보구 코드가 과하게 길어지거나 복잡해지면 기존 1번으로 유지해볼까요?

never-better commented:
좋습니다! 2번 방법 하면서 클래스나 프로토콜 사용해서 재사용하는 방법도 생각해볼 수 있으면 좋을 것 같아요.
따로 PR 파시나요 아니면 이 PR에서 계속 작업하시나요??

yangsubinn commented:
> 좋습니다! 2번 방법 하면서 클래스나 프로토콜 사용해서 재사용하는 방법도 생각해볼 수 있으면 좋을 것 같아요. 따로 PR 파시나요 아니면 이 PR에서 계속 작업하시나요??

여기서 한 서비스에서만 해보고 어떤 방식으로 할지 확정해보려구요!

yangsubinn commented:
Task+에 작성했던 retrying 함수를 아래처럼 수정했습니다! @never-better 
- retrying 내부에서 unauthorized 에러를 먼저 잡으면 jwt 토큰 재발급 서버 통신을 보낸 뒤, 다시 한 번 통신 재요청 하게 됩니다.

retrying 내에서 기존 작업을 실행하면 
- 한 번에 통신 성공하는 경우, 바로 받아온 데이터를 리턴
- 최초 통신에서 토큰 만료(unauthorized) 받는 경우, 토큰 재발급 요청 후 다시 통신 요청
- 최초 통신에서 그 외의 에러를 받는 경우, 다시 한 번 시도 후 두번째에는 에러 리턴

MeetingDetailService만 한 번 수정해봤습니다.
기존에 NetworkService의 request와 requestWithNoResult에서 실행하던 retrying을  service 내부에서 실행하고 value를 리턴하도록 했습니다!

MeetingDetailService 작업한거만 올리고 나머지 service는 새로 이슈 파서 해두겠습니다

```
extension Task where Failure == Error {
    @discardableResult
    static func retrying(
        priority: TaskPriority? = nil,
        maxRetryCount: Int = 1,
        retryDelay: TimeInterval = 0.4,
        operation: @Sendable @escaping () async throws -> Success
    ) -> Task {
        Task(priority: priority) {
            @Dependency(\.tokenRefreshService) var refreshService
            
            for _ in 0..<maxRetryCount {
                do {
                    return try await operation()
                } catch {
                    if let error = error as? APIError {
                        if error == .unauthorized {
                            if await refreshService.refresh() == .success { continue }
                        }
                    }
                    print("✅ retrying")
                    try await Task<Never, Never>.sleep(seconds: retryDelay)
                    continue
                }
            }
            
            try Task<Never, Never>.checkCancellation()
            return try await operation()
        }
    }
}

```
![스크린샷 2023-09-07 오후 10 35 55](https://github.com/dnd-side-project/dnd-9th-2-ios/assets/81167570/61666a9a-b332-4e4d-9ea0-ea25db5d1b57)

yangsubinn committed:
Fix #240 - 토큰 재발급 요청 로직 수정
- Task+ retrying 함수 내에서 토큰 재발급 요청
- NetworkService request의 retrying 함수 제거
- MeetingDetailService retrying, 토큰 재발급 로직 적용
- 불필요한 코드 삭제

never-better commented:
@yangsubinn 

오 코드 완전 깔끔하게 재사용될 것 같아요 👍

1. 만약에 토큰 리프레시 요청에서 에러가 나면 어떻게 처리되는 건가요??
2. Task안에 `@Dependency`가 필요할까요?? 저는 뭔가 `@Dependency`는 Feature 안에서만 쓰고 Task 함수 안에 TCA 의존성 없이 작성하는게 더 좋은 코드 구조라고 생각합니다. (작업하고 나중에 생각해도 되는 문제입니다)

yangsubinn commented:
> @yangsubinn 
> 
> 
> 
> 오 코드 완전 깔끔하게 재사용될 것 같아요 👍
> 
> 
> 
> 1. 만약에 토큰 리프레시 요청에서 에러가 나면 어떻게 처리되는 건가요??
> 
> 2. Task안에 `@Dependency`가 필요할까요?? 저는 뭔가 `@Dependency`는 Feature 안에서만 쓰고 Task 함수 안에 TCA 의존성 없이 작성하는게 더 좋은 코드 구조라고 생각합니다. (작업하고 나중에 생각해도 되는 문제입니다)
> 
> 

1. 현재 코드로는 최초에 토큰 리프레시 요청에 실패하면 0.4초 후 다시 기존 요청한 통신부터 재시도하게 됩니다.
토큰 요청에 실패하는 경우 바로 빼줄 수도 있겠지만, 토큰 요청도 다른 서버 요청과 동일하게 아주 잠시.. 서버 에러가 있을 수 있기 때문에 
최초 통신 실패 + 최초 토큰 요청 실패한 경우도 한번 더 처음부터 시도하도록 했습니다 !

2. 동의합니다! TCA 의존 없이 수정하게 된다면 내부에서 바로 networkService<UserAPI> 생성해서 요청보내도록 수정해야하나 싶은데 어떠신가요?

never-better commented:
1. 확인했습니다!
2. 바로 NetworkService를 생성하면 코드가 너무 길어지니까, 지금처럼 Token Refresh Service를 사용하돼 DependencyKey를 상속받지 않는 객체를 사용하면 좋을 것 같아요.

yangsubinn committed:
Fix #240 - TokenRefreshService Dependency 제거, static 함수로 수정

yangsubinn commented:
TokenRefreshService 내부 수정해서, Task+ retrying에서 TokenRefreshService.refresh() 로 사용할 수 있도록 수정했습니다.
https://github.com/dnd-side-project/dnd-9th-2-ios/pull/240/commits/db63d2f3371001dc33a8371fb2862b09091658db

+notFound일때 토큰 리프레시 요청하도록 해서 정상적으로 갱신하는 것도 확인했습니다.
![스크린샷 2023-09-08 오전 10 17 13](https://github.com/dnd-side-project/dnd-9th-2-ios/assets/81167570/e1be8b16-6b8a-489d-aaea-015e3ba17dd6)

never-better reviewed (approved):
쉽지 않았는데 고생하셨습니다 🙌

yangsubinn committed:
Merge: develop pull
Merge branch 'develop' of https://github.com/dnd-side-project/dnd-9th-2-ios into feature/jwt-renew-api
